### PR TITLE
Create: 2306 - Naming a Company (ts,js,py)

### PR DESCRIPTION
**Pull request by AHTHneeuhl**
[//]: # "Pull Request Template"
[//]: # "Replace the placeholder values in the template below"

- **File(s) Modified**: _2306-naming-a-company.(js,ts,py), etc..._
- **Language(s) Used**: _python, javascript, typescript, etc..._
- **Submission URL**: _https://leetcode.com/submissions/detail/992771955/, https://leetcode.com/submissions/detail/991962468/_

[//]: # "Getting the Submission URL"
[//]: # "Go to the leetcode [`Submissions tab`](https://user-images.githubusercontent.com/71089234/180188604-b1ecaf90-bf27-4fd6-a559-5567aebf8930.png)"
[//]: # "and [click on the `Accepted` status of your submission.](https://user-images.githubusercontent.com/71089234/180189321-1a48c33f-aa65-4b29-8aaa-685f4f5f8c9e.png)]"
[//]: # "Finally copy the URL from the nav bar, it should look like https://leetcode.com/problems/[problem-name]/submissions/xxxxxxxxx/"

### Important
Please make sure the file name is lowercase and a duplicate file does not already exist before merging.

### CODE REVIEW

File: javascript/2306-naming-a-company.js [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+ * @param {string[]} ideas
+ * @return {number}
+ */
+
+// Time complexity: O(n^2 * m)
+// Space complexity: O(n * m)
+
+var distinctNames = function (ideas) {
+    const wordMap = new Map();
+    let res = 0;
```
Comment (aakhtar3) [inline javascript/2306-naming-a-company.js:11]: Suggestion: provide a better name

count or distinctCount

File: javascript/2306-naming-a-company.js [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+
+    ideas.forEach((word) => {
+        const ch = word.charAt(0);
+        const substr = word.slice(1);
+        if (!wordMap.has(ch)) {
```
Comment (aakhtar3) [inline javascript/2306-naming-a-company.js:16]: Suggestion: leverage a one liner

```js
const set = (map.get(char) ?? new Set());

set.add(char);

map.set(char, set);
```

File: javascript/2306-naming-a-company.js [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+ */
+
+// Time complexity: O(n^2 * m)
+// Space complexity: O(n * m)
+
+var distinctNames = function (ideas) {
+    const wordMap = new Map();
+    let res = 0;
+
+    ideas.forEach((word) => {
```
Comment (aakhtar3) [inline javascript/2306-naming-a-company.js:13]: Suggestion: avoid functional for each of you can

File: javascript/2306-naming-a-company.js [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+        }
+        wordMap.get(ch).add(substr);
+    });
+
+    for (const [ch1, set1] of wordMap.entries()) {
+        for (const [ch2, set2] of wordMap.entries()) {
+            if (ch1 === ch2) {
+                continue;
+            }
+            let intersect = 0;
```
Comment (aakhtar3) [inline javascript/2306-naming-a-company.js:27]: Suggestion: move to helper function
```js
const intersect = getIntersect();
```

**Review by aakhtar3 — CHANGES_REQUESTED** [review]

File: javascript/2306-naming-a-company.js (hunk 4 above)
Comment (AHTHneeuhl) [reply javascript/2306-naming-a-company.js]: Thanks! Updated the code

Comment (aakhtar3) [reply javascript/2306-naming-a-company.js]: Few other suggestions pending

Comment (AHTHneeuhl) [reply javascript/2306-naming-a-company.js]: Updated code: now using for loop instead of forEach.

File: javascript/2306-naming-a-company.js [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+        wordMap.set(ch, set);
+    }
+
+    for (const [ch1, set1] of wordMap) {
+        for (const [ch2, set2] of wordMap) {
+            if (ch1 === ch2) {
+                continue;
+            }
+
+            const intersect = getIntersection(set1, set2);
```
Comment (aakhtar3) [inline javascript/2306-naming-a-company.js:27]: You can precompute this const and lower your time complexity

File: javascript/2306-naming-a-company.js [hunk 6]
```diff
@@ -0,0 +1,45 @@
...
+            count += distinct1 * distinct2;
+        }
+    }
+
+    return count;
+};
+
+function getIntersection(set1, set2) {
+    let intersect = 0;
+    for (const word of set1) {
```
Comment (aakhtar3) [inline javascript/2306-naming-a-company.js:39]: I suggested avoid functional looping, but in this case you can take advantage of filter when doing set intersections

```suggestion
comet getIntersection = (set1, set2) => [ ...set1 ]
    .filter((key) => set2.has(key))
    .length;
```

**Review by aakhtar3 — CHANGES_REQUESTED** [review]

File: javascript/2306-naming-a-company.js (hunk 6 above)
Comment (AHTHneeuhl) [reply javascript/2306-naming-a-company.js]: here as well updated the code.

**Pull request closed without merging**